exporter/prometheusremotewriteexporter/wal.go

```diff
@@ -0,0 +1,246 @@
+	if wc == nil {
+		return 0
+	}
+	if wc.RefreshDuration <= 0 {
+		// Following what Prometheus's WAL defaults.
+		return 1 * time.Minute
+	}
+	return wc.RefreshDuration
+}
+
+func (prwe *PrwExporter) walEnabled() bool { return prwe.walConfig != nil }
+
+var errWALDisabled = errors.New("WAL is disabled")
+
+func (prwe *PrwExporter) retrieveWALIndices(context.Context) (err error) {
+	if !prwe.walEnabled() {
+		return errWALDisabled
+	}
+
+	prwe.walMu.Lock()
+	defer prwe.walMu.Unlock()
+
+	wal, err := prwe.walConfig.createWAL()
+	if err != nil {
+		return err
+	}
+	prwe.wal = wal
+
+	prwe.rWALIndex, err = prwe.wal.FirstIndex()
+	if err != nil {
+		return fmt.Errorf("prometheusremotewriteexporter: failed to retrieve the last WAL index: %w", err)
+	}
+	prwe.wWALIndex, err = prwe.wal.LastIndex()
+	if err != nil {
+		return fmt.Errorf("prometheusremotewriteexporter: failed to retrieve the first WAL index: %w", err)
+	}
+	return nil
+}
+
+// exportThenTruncateWAL uploads the serialized data to the remote write endpoint,
+// and then truncates and retrieves the new indices for the WAL.
+func (prwe *PrwExporter) exportThenTruncateWAL(ctx context.Context, reqL []*prompb.WriteRequest) error {
+	if len(reqL) == 0 {
+		return nil
+	}
+	// Otherwise, time to flush.
+	if errs := prwe.actualExport(ctx, reqL); len(errs) != 0 {
+		return consumererror.Combine(errs)
+	}
+	// Save all the entries that aren't yet committed, to the tail of the WAL.
+	if err := prwe.wal.Sync(); err != nil {
+		return err
+	}
+	// Truncate the WAL from the front for the entries that we already
+	// read from the WAL and then exported.
+	if err := prwe.wal.TruncateFront(prwe.rWALIndex); err != nil && err != wal.ErrOutOfRange {
+		return err
+	}
+	// Reset by retrieving the respective read and write WAL indices.
+	return prwe.retrieveWALIndices(ctx)
+}
+
+func (prwe *PrwExporter) turnOnWALIfEnabled() error {
+	if !prwe.walEnabled() {
+		return nil
+	}
+
+	// From this point down below, the Write-Ahead-Log is enabled.
+	shutdownCtx, cancel := context.WithCancel(context.Background())
+	if err := prwe.retrieveWALIndices(shutdownCtx); err != nil {
+		cancel()
+		return err
+	}
+
+	go func() {
+		<-prwe.closeChan
+		cancel()
+	}()
+
+	prwe.wg.Add(1)
+	go func() {
+		defer prwe.wg.Done()
+
+		_ = prwe.readFromWALThenExport(shutdownCtx)
+	}()
+
+	return nil
+}
+
+func (prwe *PrwExporter) closeWAL() {
+	prwe.walMu.Lock()
+	defer prwe.walMu.Unlock()
+
+	if prwe.wal != nil {
+		prwe.wal.Close()
+		prwe.wal = nil
+	}
+}
+
+func (prwe *PrwExporter) writeToWAL(requests []*prompb.WriteRequest) error {
+	if !prwe.walEnabled() {
+		return errWALDisabled
+	}
+
+	// Write in batch.
+	batch := new(wal.Batch)
+	for _, req := range requests {
+		protoBlob, err := proto.Marshal(req)
+		if err != nil {
+			return err
+		}
+		prwe.wWALIndex++
+		batch.Write(prwe.wWALIndex, protoBlob)
+	}
+
+	return prwe.wal.WriteBatch(batch)
+}
+
+func (prwe *PrwExporter) readFromWALThenExport(ctx context.Context) error {
```

Comment:
How are we flushing the WAL at a Shutdown?

Reply:
On shutdown, the WAL's .Close() method is invoked and that toggles a shutdown, before invoking this method in (*PrwExporter).turnOnWALIfEnabled(), we derive a cancellable context.Context and after the prwe.closeChan receives a value, it also invokes cancel() which ensures that this method will terminate alright.